### DISCUSSION TITLE
what will happen if I frequently change the "main" branch?

### DISCUSSION THREAD

**Discussion by xxleyi** [arxanas/git-branchless#1201] [Q&A]
### Wiki page

https://github.com/arxanas/git-branchless/wiki/Concepts#main-branch

### Subcommand

init

### Description

as the title say

**Comment by arxanas** [comment]
Sorry, I don't understand the question. Why would something significant happen if you frequently change the main branch? Is there a particular part of the documentation you're referring to?

**Reply by xxleyi** [reply]
> The branchless workflow assumes that all of your work is based off of a single main branch. It's not able to render multiple long-lived branches well.

it's this part, I am thinking if I can use git branchless if I have multiple long-lived branches by frequently changing main branch

**Reply by arxanas** [reply]
The problems with multiple long-lived branches are:

- They don't render well in the smartlog by default. However, you can adjust this setting to exclude some of the branches/extra commits: https://github.com/arxanas/git-branchless/wiki/Configuration#branchlesssmartlogdefaultrevset
  - This option did not exist when that section was originally written.
- Some of the built-in revsets (like `draft()`/`stack()`) may not return the results you expect, as they are relative to the main branch. Likewise, you may have to avoid `git sync` or pass your own revset. But those should be pretty easy to work around.
- If you have a *lot* of branches, then performance on each command might suffer slightly, as git-branchless walks all branches on startup to register any new commits.

I don't think you'll have issues other than that.

**Reply by rleconte** [reply]
I came upon this thread when having the same kind of "problem", and modifying the default `revset` for `git sl` indeed works, thanks!
For information the modified `revset` looks like:
`((draft() | branches() | @) % (main() | long-living-branch)) | branches() | @`
And the result:
```
❯ git sl
⋮
◇ e105f91 30d Long living branch has been branched here
┣━┓
⋮ ⊘ 76 omitted commits
⋮ ⋮
⋮ ◯ 074b099 2d (long-living-branch) Last commit on long living branch
⋮ ┃
⋮ ◯ 25b4d2a 1m Some commit on a branch of long living branch
⋮ ┃
⋮ ◯ a3a8dfd 1m (branch-on-long-living-branch) Last commit on my branch
⋮
◆ 6518cd8 1d (ᐅ main) Last commit in main
┃
◯ e63aab1 22s (branch-on-main) Some work on main
```

As I feel it is a common workflow in enterprises to have release branches, that are as "hard as a diamond" as `main`, should I open a feature request to declare other branches that could be treated as main, or it has already been proposed and rejected for some reason?

The expected result of such a feature would be to have a variation of the following smartlog:
```
❯ git sl
⋮
◇ e105f91 30d Long living branch has been branched here
┣━┓
⋮ ◆ 074b099 2d (long-living-branch) Last commit on long living branch
⋮ ┃
⋮ ◯ 25b4d2a 1m Some commit on a branch of long living branch
⋮ ┃
⋮ ◯ a3a8dfd 1m (branch-on-long-living-branch) Last commit on my branch
⋮
◆ 6518cd8 1d (ᐅ main) Last commit in main
┃
◯ e63aab1 22s (branch-on-main) Some work on main
```

**Reply by arxanas** [reply]
The main problem is that commands like [`git sync`] and revsets like `stack()` (and, of course, `main()`) are defined in terms of the main branch. In your workflow, does declaring multiple main branches do anything other than remove the "omitted" node from the smartlog?

**Reply by rleconte** [reply]
It does!
I also adapted `git sync` to my uses, with an alias:
`git sync 'draft() - :release-branch - descendants(release-branch)'`

Then it allows me to call this alias to sync all my branches and commits on top of `main` but also `release-branch` (for fixes that need to go on the release)

So I don't use all the commands yet, and don't know this project well enough to judge if having `main()` revset reference main + release branches would break things in all kind of weird ways, but so far I had to adapt `git sl` and `git sync` to treat those branches as immutable stuff on top of which I'm contributing commits and branches.

**Reply by arxanas** [reply]
@rleconte In the general case, if a commit to be synced has a common merge-base with multiple long-lived branches, how will we know which one to rebase it on top of?

**Reply by rleconte** [reply]
That's a good question, that I did not ask myself because with the way our branches are set up, once `release-branch` has been branched off, it never "crosses" `main` again, everything that needs to go in `release-branch` is cherry picked, not merged into it.
So I guess that's why my alias `git sync 'draft() - :release-branch - descendants(release-branch)'` that I talked about earlier works the way I expect it to (meaning a bunch of commits on `release-branch` are rebased on top of the newer version of `release-branch` and the same for `main`).
